pkgs/applications/misc/hello/salve-mundi.nix

```diff
@@ -0,0 +1,98 @@
+/*
+  This file showcases various techniques and features of Nixpkgs, whereas the
+  regular hello serves as more of an introductory package definition.
+  Salve mundi is latin for hello world.
+*/
+{ callPackage
+, lib
+, stdenv
+, fetchurl
+, nixos
+, testers
+, salve-mundi
+, makeWrapper
+, runCommand
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "salve-mundi";
+  version = "2.12.1";
+
+  src = fetchurl {
+    url = "mirror://gnu/hello/hello-${finalAttrs.version}.tar.gz";
+    sha256 = "sha256-jZkUKv2SV28wsM18tCqNxoCZmLxdYH2Idh9RLibH2yA=";
+  };
+
+  doCheck = true;
+
+
+  /* Wrap the program */
+
+  postInstall = lib.optionalString (finalAttrs.passthru.greeting != null) ''
+    wrapProgram $out/bin/hello --append-flags --greeting=${lib.escapeShellArg (lib.escapeShellArg finalAttrs.passthru.greeting)}
+  '';
+  nativeBuildInputs = [ makeWrapper ];
+  passthru.greeting = "SALVE MUNDI";
+
+
+  /* Tidy up the package attributes and make them useful */
+
+  __cleanAttrs = true;
+  passthru.exe = lib.getExe finalAttrs.finalPackage;
+  # We allow `hello.src` to be used in tests and examples, despite __cleanAttrs
+  passthru.src = finalAttrs.src;
```

Comment:
IMO src should always be passed through. eg you sometimes want to (re-)build it and as I understand it, that would after that no longer be possible which would suck.

Reply:
> For me this translates into more work and discussions. Sometimes I know that the solution I am drafting is hacky and I decided for myself that it is better than the available alternatives. I don't want to contact another person to expose some variable(s) for that though.

You'll still be able to do this, by using attributes such as `.internals.src`. If you don't think your use case should be explicitly supported by the package, you don't need to do anything.

Also if you're very confident about your use case, you could just open a PR. It doesn't have to some sort of drawn out discussion.

> I personally would want a setting to just disable cleanAttrs everywhere and give me access to everything, as I don't really care if something is marked as an interface or not. Or do we now need to start doing reflections?

What do you mean with reflections?

I'm not in favor of such a global flag, because it is not necessary and it will make some expressions that do or do not use the flag incompatible. (Configuration should always be a last resort)

> For example this would majorly break nix-update and complicate it.
> 
> https://github.com/Mic92/nix-update/blob/f7eba589cc99a0cb6f0bab298e194c7c5705be13/nix_update/eval.py#L106-L118

It'd be fine for `nix-update` to peek inside the `internals` and then carry on as it did.

It would be nicer if the language integrations and `nix-update` agreed on some small but general interface, such as a package attribute, so that `nix-update` isn't coupled to each of the language integrations. More concretely, if the language integrations provided a default `updateScript` attribute, `nix-update` would be trivial. That's a separate project though that I don't really want to delve into here, but I think it shows that `internals` highlights the places where the design could be improved.

Reply:
`src` -> `internals.src` should be fine for individual "override" packages you write.

There is no general way to get all sources in using just the expression language, so we don't lose the capability to generically retrieve all sources, because we never had that in the language. Simply put, not all packages have a `src`, and not all sources appear in `src`, and you can hide "anything" in a dependency of a derivation a string context.
This task is better achieved by analyzing the `.drv` graph, you can find _all_ the derivation input sources (file references and builtin fetcher results) and fixed output derivations. You may even do this from within a derivation.